### PR TITLE
feat: Amazon SageMaker compatible images

### PR DESCRIPTION
**Pull request by JGalego**
# What does this PR do?

Adds support for [Amazon SageMaker](https://aws.amazon.com/sagemaker/) compatible images. Similar to https://github.com/huggingface/text-generation-inference/pull/147, only for TEI.
Mostly CI stuff and some [hacks](https://github.com/huggingface/text-generation-inference/pull/147/commits/930842a7f0ed6ea26211c8bb2f3dbc596f1afde5), since the [Amazon SageMaker routes](https://github.com/huggingface/text-embeddings-inference/blob/282812743444c33f9e5f4f3681dbbe2472fd651e/router/src/http/server.rs#L814) are already implemented.

**Implementation:**
+ Added `sagemaker` target to `Dockerfile-cuda` and custom entrypoint
+ Added `build-and-push-sagemaker-image` step to `build_*` workflows

## Who can review?

@OlivierDehaene

### CODE REVIEW

**Comment by austinmw** [comment]
Hi, happen to know if this supports `BAAI/bge-reranker-base`?

**Comment by JGalego** [comment]
> Hi, happen to know if this supports `BAAI/bge-reranker-base`?

It does. Running the same example as in https://github.com/huggingface/text-embeddings-inference#using-re-rankers-models
image

**Comment by superlaut** [comment]
Hi team, is there an expected date for merging this PR? I would like to use these images :)

**Comment by andrewrreed** [comment]
FYI - here's a working notebook to deploy TEI on SageMaker in the meantime: https://github.com/andrewrreed/hf-notebooks/blob/main/deploy-tei-sagemaker/tei-sagemaker-inference.ipynb

**Comment by philschmid** [comment]
Hey @JGalego,

Thank you for starting this work. We would need to make some changes to match the environment variables, similar to TGI. https://github.com/huggingface/text-generation-inference/blob/main/sagemaker-entrypoint.sh
Do you want to make those changes? If you are busy we can also contribute directly into your branch. 

Also due to complexity of SageMaker, we will build 1 GPU container containing all binaries for the target (https://github.com/huggingface/text-embeddings-inference/blob/main/Dockerfile-cuda-all). 
Could you make the changes to the `dockerfile-cuda-all`?

File: .github/workflows/build_75.yaml
Comment (philschmid) [inline .github/workflows/build_75.yaml:1]: Lets update the workflow file only for `all`. https://github.com/huggingface/text-embeddings-inference/blob/main/.github/workflows/build_all.yaml

File: Dockerfile-cuda [hunk 1]
```diff
@@ -108,6 +108,16 @@ ENV HUGGINGFACE_HUB_CACHE=/data \
     PORT=80 \
     USE_FLASH_ATTENTION=$DEFAULT_USE_FLASH_ATTENTION
 
+# Amazon SageMaker compatible image
```
Comment (philschmid) [inline Dockerfile-cuda:111]: Lets put this also only in all https://github.com/huggingface/text-embeddings-inference/blob/main/Dockerfile-cuda-all

**Review by philschmid — COMMENTED** [review]
Lets start only with the "all" versions.

File: Dockerfile-cuda (hunk 1 above)
Comment (JGalego) [reply Dockerfile-cuda]: Moved `sagemaker` target to `Dockerfile-cuda-all` and updated `sagemaker-entrypoint.sh`. 

Ran into this issue https://github.com/huggingface/candle/issues/733 while running some tests on `g5.2xlarge` endpoints - added a `CUDA_COMPUTE_CAP` env var as a workaround.

image

File: .github/workflows/build_75.yaml
Comment (JGalego) [reply .github/workflows/build_75.yaml:1]: Reverted changes to workflow files and added build+push step for sagemaker to the `all` workflow

**Review by philschmid — APPROVED** [review]
Thank you for working on this! As soon as this is merged we ll work with the SageMaker to make it officially available in side the `sagemaker-sdk`.

**Review by OlivierDehaene — APPROVED** [review]
Thank you for this and sorry for the delay!

**Comment by austinmw** [comment]
@JGalego  Hi, I have a question. Does this work when using an S3 path for model artifacts instead of HF Hub, as in `HuggingFaceModel(model_data="<s3_path>", ...)`?

I get an endpoint error if I don't specify `HF_MODEL_ID` as an env config, however I thought that `model_data` will be ignored if I specify this?

Appreciate any advice, thanks!

**Comment by JGalego** [comment]
> @JGalego Hi, I have a question. Does this work when using an S3 path for model artifacts instead of HF Hub, as in `HuggingFaceModel(model_data="<s3_path>", ...)`?
> 
> I get an endpoint error if I don't specify `HF_MODEL_ID` as an env config, however I thought that `model_data` will be ignored if I specify this?
> 
> Appreciate any advice, thanks!

For the SM images, `HF_MODEL_ID` is **mandatory** (https://github.com/huggingface/text-embeddings-inference/blob/main/sagemaker-entrypoint-cuda-all.sh#L25). Checking the code (https://github.com/huggingface/text-embeddings-inference/blob/main/router/src/lib.rs#L95) and the docs (https://github.com/huggingface/text-embeddings-inference?tab=readme-ov-file#docker), it only supports model artifacts from the hub.

```
Options:
      --model-id <MODEL_ID>
          The name of the model to load. Can be a MODEL_ID as listed on <https://hf.co/models> like `thenlper/gte-base`.
          Or it can be a local directory containing the necessary files as saved by `save_pretrained(...)` methods of
          transformers

          [env: MODEL_ID=]
          [default: thenlper/gte-base]
```

**Comment by philschmid** [comment]
Hello @austinmw @JGalego,

This works already. Just set the `HF_MODEL_ID` to `/opt/ml/model`. SageMaker loads your model from s3 to `/opt/ml/model` and TEI starts correctly.

**Comment by philschmid** [comment]
@austinmw This looks more that your artifact is not correctly created. Can you check https://huggingface.co/docs/sagemaker/inference#create-a-model-artifact-for-deployment

**Comment by austinmw** [comment]
@philschmid Just tried with new artifacts using the provided steps and got the same exact error:

```
# delete my fine-tuned model from s3
aws s3 rm s3://sagemaker-us-east-1-197237853439/models/bge-base-en-v1.5/model.tar.gz

# upload fresh artifacts according to steps
git lfs install
git clone https://huggingface.co/BAAI/bge-base-en-v1.5
cd bge-base-en-v1.5/
tar zcvf model.tar.gz *
aws s3 cp model.tar.gz s3://sagemaker-us-east-1-197237853439/models/bge-base-en-v1.5/model.tar.gz
```

image

Is it possible this feature is broken for TEI?

**Comment by philschmid** [comment]
Nope it should work. I am working on a SageMaker end-to-end example with training and fine-tuning and it works. Can you try uploading your model unzipped and then deploy it using the "uncompressed" setting? 

s3_path="s3://path/to/ymodel" 

```python
model = HuggingFaceModel(
  role=role,
  # path to s3 bucket with model, we are not using a compressed model
  model_data={'S3DataSource':{'S3Uri': s3_path + "/",'S3DataType': 'S3Prefix','CompressionType': 'None'}},
  image_uri=image,
  env=config
)```

**Comment by austinmw** [comment]
That worked, thanks a ton for your help!

**Pull request merged**